File: results/2026-05-06_173435/tasks/16-environment-matrix-generator/typescript-bun-opus/generated-code/matrix-generator.ts
import type { MatrixConfig, MatrixEntry, MatrixOutput } from "./types";
import { readFileSync } from "fs";

const DEFAULT_MAX_COMBINATIONS = 256;

function validateConfig(config: unknown): MatrixConfig {
  if (typeof config !== "object" || config === null) {
    throw new Error("Config must be a JSON object");
  }
  const c = config as Record<string, unknown>;
  if (!c.matrix || typeof c.matrix !== "object") {
    throw new Error("Config must have a 'matrix' object with at least one dimension");
  }
  const matrix = c.matrix as Record<string, unknown>;
  for (const [key, values] of Object.entries(matrix)) {
    if (!Array.isArray(values) || values.length === 0) {
      throw new Error(`Dimension '${key}' must be a non-empty array`);
    }
  }
  if (c.include !== undefined && !Array.isArray(c.include)) {
    throw new Error("'include' must be an array of objects");
  }
  if (c.exclude !== undefined && !Array.isArray(c.exclude)) {
    throw new Error("'exclude' must be an array of objects");
  }
  if (c["max-parallel"] !== undefined && (typeof c["max-parallel"] !== "number" || c["max-parallel"] < 1)) {
    throw new Error("'max-parallel' must be a positive integer");
  }
  if (c["max-combinations"] !== undefined && (typeof c["max-combinations"] !== "number" || c["max-combinations"] < 1)) {
    throw new Error("'max-combinations' must be a positive integer");
  }
  return config as MatrixConfig;
}

function cartesianProduct(matrix: Record<string, (string | number | boolean)[]>): MatrixEntry[] {
  const keys = Object.keys(matrix);
  if (keys.length === 0) return [{}];

  const result: MatrixEntry[] = [];
  const values = keys.map((k) => matrix[k]);

  function recurse(depth: number, current: MatrixEntry): void {
    if (depth === keys.length) {
      result.push({ ...current });
      return;
    }
    for (const val of values[depth]) {
      current[keys[depth]] = val;
      recurse(depth + 1, current);
    }
  }
  recurse(0, {});
  return result;
}

function matchesRule(entry: MatrixEntry, rule: Record<string, string | number | boolean>): boolean {
  return Object.entries(rule).every(([key, val]) => entry[key] === val);
}

function applyExcludes(combinations: MatrixEntry[], excludes: Record<string, string | number | boolean>[]): MatrixEntry[] {
  return combinations.filter((entry) => !excludes.some((rule) => matchesRule(entry, rule)));
}

function applyIncludes(
  combinations: MatrixEntry[],
  includes: Record<string, string | number | boolean>[],
  dimensionKeys: string[]
): MatrixEntry[] {
  const result = combinations.map((e) => ({ ...e }));

  for (const inc of includes) {
    const matchIdx = result.findIndex((entry) =>
      dimensionKeys.every((k) => !(k in inc) || entry[k] === inc[k])
    );

    const hasNewDimensionValue = dimensionKeys.some(
      (k) => k in inc && !combinations.some((e) => e[k] === inc[k])
    );

    if (hasNewDimensionValue || matchIdx === -1) {
      result.push({ ...inc });
    } else {
      Object.assign(result[matchIdx], inc);
    }
  }
  return result;
}

export function generateMatrix(config: MatrixConfig): MatrixOutput {
  const validated = validateConfig(config);
  const maxCombinations = validated["max-combinations"] ?? DEFAULT_MAX_COMBINATIONS;
  const dimensionKeys = Object.keys(validated.matrix);

  let combinations = cartesianProduct(validated.matrix);

  if (validated.exclude) {
    combinations = applyExcludes(combinations, validated.exclude);
  }

  if (validated.include) {
    combinations = applyIncludes(combinations, validated.include, dimensionKeys);
  }

  if (combinations.length > maxCombinations) {
    throw new Error(
      `Matrix produces ${combinations.length} combinations, exceeding limit of ${maxCombinations}`
    );
  }

  if (combinations.length === 0) {
    throw new Error("Matrix produces 0 combinations after applying include/exclude rules");
  }

  const dimensions: Record<string, number> = {};
  for (const key of dimensionKeys) {
    dimensions[key] = validated.matrix[key].length;
  }

  const output: MatrixOutput = {
    strategy: {
      "fail-fast": validated["fail-fast"] ?? true,
      matrix: {
        include: combinations,
      },
    },
    total_combinations: combinations.length,
    dimensions,
  };

  if (validated["max-parallel"] !== undefined) {
    output.strategy["max-parallel"] = validated["max-parallel"];
  }

  return output;
}

if (import.meta.main) {
  const args = process.argv.slice(2);
  if (args.length !== 1) {
    console.error("Usage: bun run matrix-generator.ts <config.json>");
    process.exit(1);
  }

  try {
    const raw = readFileSync(args[0], "utf-8");
    const config = JSON.parse(raw);
    const result = generateMatrix(config);
    console.log(JSON.stringify(result, null, 2));
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`ERROR: ${msg}`);
    process.exit(1);
  }
}

File: results/2026-05-06_173435/tasks/16-environment-matrix-generator/default-opus/generated-code/matrix_generator.py
#!/usr/bin/env python3
"""Environment Matrix Generator for GitHub Actions.

Generates a build matrix JSON suitable for GitHub Actions strategy.matrix.
Supports include/exclude rules, max-parallel limits, fail-fast, and
validates that the matrix doesn't exceed a configurable maximum size.
"""

import json
import sys
import itertools

DEFAULT_MAX_COMBINATIONS = 256
DEFAULT_FAIL_FAST = False


def validate_config(config: dict) -> None:
    """Validate the matrix configuration structure."""
    if not isinstance(config, dict):
        raise ValueError("Configuration must be a JSON object")

    if "matrix" not in config:
        raise ValueError("Configuration must contain a 'matrix' key with dimension definitions")

    matrix = config["matrix"]
    if not isinstance(matrix, dict):
        raise ValueError("'matrix' must be a JSON object mapping dimension names to value arrays")

    if not matrix:
        raise ValueError("'matrix' must contain at least one dimension")

    for key, values in matrix.items():
        if not isinstance(values, list):
            raise ValueError(f"Dimension '{key}' must be an array of values, got {type(values).__name__}")
        if not values:
            raise ValueError(f"Dimension '{key}' must have at least one value")

    for rule_key in ("include", "exclude"):
        if rule_key in config:
            rules = config[rule_key]
            if not isinstance(rules, list):
                raise ValueError(f"'{rule_key}' must be an array of objects")
            for i, rule in enumerate(rules):
                if not isinstance(rule, dict):
                    raise ValueError(f"'{rule_key}[{i}]' must be a JSON object")

    if "max-parallel" in config:
        mp = config["max-parallel"]
        if not isinstance(mp, int) or mp < 1:
            raise ValueError("'max-parallel' must be a positive integer")

    if "max-combinations" in config:
        mc = config["max-combinations"]
        if not isinstance(mc, int) or mc < 1:
            raise ValueError("'max-combinations' must be a positive integer")


def compute_cross_product(matrix: dict) -> list[dict]:
    """Compute the Cartesian product of all matrix dimensions."""
    keys = sorted(matrix.keys())
    values = [matrix[k] for k in keys]
    return [dict(zip(keys, combo)) for combo in itertools.product(*values)]


def matches_rule(combination: dict, rule: dict) -> bool:
    """Check if a combination matches all key-value pairs in a rule."""
    return all(combination.get(k) == v for k, v in rule.items())


def apply_excludes(combinations: list[dict], excludes: list[dict]) -> list[dict]:
    """Remove combinations matching any exclude rule."""
    return [c for c in combinations
            if not any(matches_rule(c, rule) for rule in excludes)]


def apply_includes(combinations: list[dict], includes: list[dict], matrix_keys: set) -> list[dict]:
    """Apply include rules following GitHub Actions semantics.

    If an include specifies only existing matrix keys, it adds a new combination
    (if not already present). If it specifies new keys, those are merged into
    matching existing combinations; if nothing matches, a new combination is added.
    """
    result = [dict(c) for c in combinations]

    for include in includes:
        existing_keys = {k: v for k, v in include.items() if k in matrix_keys}
        extra_keys = {k: v for k, v in include.items() if k not in matrix_keys}

        if extra_keys:
            matched = False
            for combo in result:
                if matches_rule(combo, existing_keys):
                    combo.update(extra_keys)
                    matched = True
            if not matched:
                result.append(dict(include))
        else:
            already_exists = any(
                all(combo.get(k) == v for k, v in include.items())
                for combo in result
            )
            if not already_exists:
                result.append(dict(include))

    return result


def count_effective_combinations(config: dict) -> int:
    """Count effective combinations after applying include/exclude rules."""
    matrix = config["matrix"]
    combinations = compute_cross_product(matrix)

    excludes = config.get("exclude", [])
    if excludes:
        combinations = apply_excludes(combinations, excludes)

    includes = config.get("include", [])
    if includes:
        combinations = apply_includes(combinations, includes, set(matrix.keys()))

    return len(combinations)


def generate_matrix(config: dict) -> dict:
    """Generate the complete strategy configuration for GitHub Actions.

    Returns a dict with matrix, fail-fast, max-parallel, and total_combinations.
    Raises ValueError if the matrix exceeds max-combinations.
    """
    validate_config(config)

    max_combinations = config.get("max-combinations", DEFAULT_MAX_COMBINATIONS)
    total = count_effective_combinations(config)

    if total > max_combinations:
        raise ValueError(
            f"Matrix produces {total} combinations, exceeding the maximum of "
            f"{max_combinations}. Add exclude rules or reduce dimension values."
        )

    output_matrix = {}
    for key, values in sorted(config["matrix"].items()):
        output_matrix[key] = list(values)

    if config.get("include"):
        output_matrix["include"] = config["include"]

    if config.get("exclude"):
        output_matrix["exclude"] = config["exclude"]

    strategy = {
        "matrix": output_matrix,
        "fail-fast": config.get("fail-fast", DEFAULT_FAIL_FAST),
        "total_combinations": total,
    }

    if "max-parallel" in config:
        strategy["max-parallel"] = config["max-parallel"]

    return strategy


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 matrix_generator.py <config.json>", file=sys.stderr)
        sys.exit(1)

    config_path = sys.argv[1]

    try:
        with open(config_path) as f:
            config = json.load(f)
        result = generate_matrix(config)
        print(json.dumps(result, indent=2, sort_keys=True))
    except FileNotFoundError:
        print(f"Error: Config file not found: {config_path}", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)
    except ValueError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
